src/NServiceBus.TransportTests/When_multiple_messages_are_available.cs

```diff
@@ -0,0 +1,62 @@
+﻿namespace NServiceBus.TransportTests;
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using NUnit.Framework;
+using Transport;
+
+public class When_multiple_messages_are_available : NServiceBusTransportTest
+{
+    [TestCase(TransportTransactionMode.None)]
+    [TestCase(TransportTransactionMode.ReceiveOnly)]
+    [TestCase(TransportTransactionMode.SendsAtomicWithReceive)]
+    [TestCase(TransportTransactionMode.TransactionScope)]
+    public async Task Should_invoke_on_message_in_parallel(TransportTransactionMode transactionMode)
+    {
+        const int concurrencyLevel = 10;
+        var onMessageCalls = new ConcurrentQueue<TaskCompletionSource>();
+
+        await StartPump(async (context, _) =>
+            {
+                var tcs = CreateTaskCompletionSource();
+                onMessageCalls.Enqueue(tcs);
+                // "block" current pipeline invocation
+                await tcs.Task;
+            },
+            (errorContext, __) => throw new Exception("unexpected error", errorContext.Exception),
+            transactionMode,
+            pushRuntimeSettings: new PushRuntimeSettings(concurrencyLevel));
+
+        for (int i = 0; i < concurrencyLevel * 2; i++)
+        {
+            await SendMessage(InputQueueName);
+        }
+
+        TestContext.WriteLine("before assertion");
+
+        TestContext.WriteLine($"before await {DateTime.UtcNow:F}");
+        // we need to wait because it might take a bit till the pump has invoked all pipelines?
+        while (onMessageCalls.Count < concurrencyLevel)
+        {
+            await Task.Delay(50, TestTimeoutCancellationToken);
+        }
+
+        Assert.AreEqual(concurrencyLevel, onMessageCalls.Count, "should not process more messages than configured at once");
```

Comment:
Could we move the assert down to the other assert by storing the current count in a variable so that we have all asserts together?